File: skysail.server.queryfilter/src/io/skysail/server/queryfilter/nodes/BranchNode.java
package io.skysail.server.queryfilter.nodes;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

import io.skysail.server.domain.jvm.FieldFacet;
import io.skysail.server.filter.ExprNode;
import io.skysail.server.filter.Operation;
import lombok.Getter;

/**
 * A BranchNode is an ExprNode with children.
 *
 * The associated operation (like "AND") is applied to all children. "isLeaf"
 * will always return false.
 *
 */
public abstract class BranchNode extends AbstractExprNode {

    @Getter
    protected List<ExprNode> childList = new ArrayList<>();

    protected BranchNode(Operation op) {
        super(op);
    }

    public BranchNode(Operation op, List<ExprNode> childList) {
        super(op);
        this.childList = childList;
    }

    @Override
    public final boolean isLeaf() {
        return false;
    }

    @Override
    public Set<String> getSelected(FieldFacet facet, Map<String, String> lines) {
        Set<String> result = new HashSet<>();

        lines.keySet().forEach(key -> {
            if (lines.get(key).equals(asLdapString())) { // "(betrag<0.0)"; "(|(betrag<0.0)(&(betrag>0.0)(betrag<100.0)))"
                result.add(key);
            }
        });


        for (ExprNode child : childList) {
           result.addAll(child.getSelected(facet, lines));
        }
        return result;
    }

    @Override
    public ExprNode reduce(String value, FieldFacet facet,String format) {
        return this;
    }

    @Override
    public Set<String> getKeys() {
        Set<String> result = new HashSet<>();
        for (ExprNode exprNode : childList) {
            result.addAll(exprNode.getKeys());
        }
        return result;
    }

    @Override
    public boolean evaluateValue(Object gotten) {
        // TODO Auto-generated method stub
        return false;
    }



}

File: skysail.server.http.query/src/io/skysail/server/http/query/nodes/BranchNode.scala
package io.skysail.server.http.query.nodes

import io.skysail.server.filter.Operation
import io.skysail.server.domain.jvm.FieldFacet

import java.util.Collections;
import io.skysail.server.filter.ExprNode
import io.skysail.server.http.query.AbstractExprNode

abstract class BranchNode(op: Operation, attribute: String, value: String) extends AbstractExprNode(op) {

   def this(op: Operation, attribute: String) {
    this(op, attribute, null);
  }

  def getSelected(x$1: FieldFacet, x$2: java.util.Map[String, String]): java.util.Set[String] = {
     null
   }

  def isLeaf() = false 

  def reduce(x$1: String, x$2: FieldFacet, x$3: String): ExprNode = {
     ???
   }

  def getKeys(): java.util.Set[String] = {
     var result = new java.util.HashSet[String]();
//     for (ExprNode exprNode : childList) {
//         result.addAll(exprNode.getKeys());
//     }
     result
  }

  def evaluateValue(x$1: Any): Boolean = {
     ???
   }
}